src/cascadia/TerminalSettingsModel/ActionMap.cpp

```diff
@@ -0,0 +1,329 @@
+        for (const auto& parent : _parents)
+        {
+            parent->_PopulateNameMap(nameMap, visitedActionIDs);
+        }
+
+        // Add NestedCommands to NameMap _after_ we handle our parents.
+        // This allows us to override whatever our parents tell us.
+        for (const auto& [name, cmd] : _NestedCommands)
+        {
+            nameMap.insert_or_assign(name, cmd);
+        }
+    }
+
+    com_ptr<ActionMap> ActionMap::Copy() const
+    {
+        auto actionMap{ make_self<ActionMap>() };
+
+        std::for_each(_KeyMap.begin(), _KeyMap.end(), [actionMap](const auto& pair) {
+            actionMap->_KeyMap.insert(pair);
+        });
+
+        std::for_each(_ActionMap.begin(), _ActionMap.end(), [actionMap](const auto& pair) {
+            const auto& cmdCopy{ get_self<Command>(pair.second)->Copy() };
+            actionMap->_ActionMap.insert({ pair.first, *cmdCopy });
+        });
+        return actionMap;
+    }
+
+    // Method Description:
+    // - Adds a command to the ActionMap
+    // Arguments:
+    // - cmd: the command we're adding
+    void ActionMap::AddAction(const Model::Command& cmd)
+    {
+        // _Never_ add null to the ActionMap
+        if (!cmd)
+        {
+            return;
+        }
+
+        // _Always_ add nested commands
+        if (cmd.HasNestedCommands())
+        {
+            // But check if it actually has a name to bind to first
+            const auto name{ cmd.Name() };
+            if (!name.empty())
+            {
+                _NestedCommands.Insert(name, cmd);
+            }
+            return;
+        }
+
+        // General Case:
+        //  Add the new command to the NameMap and KeyMap (whichever is applicable).
+        //  These maps direct you to an entry in the ActionMap.
+
+        // Removing Actions from the Command Palette:
+        //  cmd.Name and cmd.Action have a one-to-one relationship.
+        //  If cmd.Name is empty, we must retrieve the old name and remove it.
+
+        // Removing Key Bindings:
+        //  cmd.Keys and cmd.Action have a many-to-one relationship.
+        //  If cmd.Keys is empty, we don't care.
+        //  If action is "unbound"/"invalid", you're explicitly unbinding the provided cmd.keys.
+
+        Model::Command oldCmd{ nullptr };
+        const auto actionID{ ActionHash()(cmd.Action()) };
+        auto actionPair{ _ActionMap.find(actionID) };
+        if (actionPair != _ActionMap.end())
+        {
+            // We're adding an action that already exists.
+            // Record it and update it with any new information.
+            oldCmd = actionPair->second;
+
+            // Update Name
+            const auto cmdImpl{ get_self<Command>(cmd) };
+            if (cmdImpl->HasName())
+            {
+                // This command has a name, check if it's new.
+                const auto newName{ cmd.Name() };
+                if (newName != oldCmd.Name())

```

Comment:
So to be sure - this is comparing `newName` (which definitely is a given name, because `HasName` will be false for a generated name) with the current name for this command - which might be a generated name (if `oldCmd` doesn't have a name explicitly set)? 

If I do something like 

```
{ "command": { "action": "newTab" } },
{ "command": { "action": "newTab" }, "name": "foo" },
{ "command": { "action": "newTab" }, "name": "bar" },
```

will I end up with one command in the palette, or 3?

Reply:
One: "bar". `ActionMap` will store the command as if it was:
```js
{ "command": { "action": "newTab" }, "name": "bar" },
```

This will be _super_ useful when I start working on serialization because we won't need to write any of the older names that got overwritten.